tests/e2e/tidbcluster/tidbcluster.go

```diff
@@ -1574,6 +1573,366 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
+			err := controller.GuaranteedUpdate(genericCli, tc, func() error {
+				pdCfg := v1alpha1.NewPDConfig()
+				tikvCfg := v1alpha1.NewTiKVConfig()
+				tidbCfg := v1alpha1.NewTiDBConfig()
+				pdCfg.Set("lease", 3)
+				tikvCfg.Set("status-thread-pool-size", 1)
+				tidbCfg.Set("token-limit", 10000)
+				tc.Spec.PD.Config = pdCfg
+				tc.Spec.TiKV.Config = tikvCfg
+				tc.Spec.TiDB.Config = tidbCfg
+				return nil
+			})
+			framework.ExpectNoError(err, "failed to update components configuration")
+
+			ginkgo.By("Wait for PD to be in UpgradePhase")
+			utiltc.MustWaitForPDPhase(cli, tc, v1alpha1.UpgradePhase, 3*time.Minute, 10*time.Second)
+			log.Logf("PD is in UpgradePhase")
+
+			ginkgo.By("Wait for tc ready")
+			err = oa.WaitForTidbClusterReady(tc, 10*time.Minute, 10*time.Second)
+			framework.ExpectNoError(err, "failed to wait for TidbCluster %s/%s components ready", ns, tc.Name)
+
+			ginkgo.By("Check PD configuration")
+			pdMemberName := controller.PDMemberName(tc.Name)
+			pdSts, err := stsGetter.StatefulSets(ns).Get(pdMemberName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get StatefulSet %s/%s", ns, pdMemberName)
+			pdCmName := member.FindConfigMapVolume(&pdSts.Spec.Template.Spec, func(name string) bool {
+				return strings.HasPrefix(name, controller.PDMemberName(tc.Name))
+			})
+			pdCm, err := c.CoreV1().ConfigMaps(ns).Get(pdCmName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get ConfigMap %s/%s", ns, pdCm)
+			log.Logf("PD config:\n%s", pdCm.Data["config-file"])
+			gomega.Expect(pdCm.Data["config-file"]).To(gomega.ContainSubstring("lease = 3"))
+
+			ginkgo.By("Wait for TiKV to be in UpgradePhase")
+			utiltc.MustWaitForTiKVPhase(cli, tc, v1alpha1.UpgradePhase, 3*time.Minute, 10*time.Second)
+			log.Logf("TiKV is in UpgradePhase")
+
+			ginkgo.By("Wait for tc ready")
+			err = oa.WaitForTidbClusterReady(tc, 10*time.Minute, 10*time.Second)
+			framework.ExpectNoError(err, "failed to wait for TidbCluster %s/%s components ready", ns, tc.Name)
+
+			ginkgo.By("Check TiKV configuration")
+			tikvMemberName := controller.TiKVMemberName(tc.Name)
+			tikvSts, err := stsGetter.StatefulSets(ns).Get(tikvMemberName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get StatefulSet %s/%s", ns, tikvMemberName)
+			tikvCmName := member.FindConfigMapVolume(&tikvSts.Spec.Template.Spec, func(name string) bool {
+				return strings.HasPrefix(name, controller.TiKVMemberName(tc.Name))
+			})
+			tikvCm, err := c.CoreV1().ConfigMaps(ns).Get(tikvCmName, metav1.GetOptions{})
+			framework.ExpectNoError(err, "failed to get ConfigMap %s/%s", ns, tikvCmName)
+			log.Logf("TiKV config:\n%s", tikvCm.Data["config-file"])
+			gomega.Expect(tikvCm.Data["config-file"]).To(gomega.ContainSubstring("status-thread-pool-size = 1"))
+
+			ginkgo.By("Wait for TiDB to be in UpgradePhase")
+			utiltc.MustWaitForTiDBPhase(cli, tc, v1alpha1.UpgradePhase, 3*time.Minute, 10*time.Second)
```

Comment:
Can we always observe these intermediate phases?

Reply:
@csuzhangxc I think we can as long as the interval is small enough.
Do you have any better solution of waiting for the correct condition?

Reply:
OK. No better solution now.